Enhance DataTable multiple pinning

leossantiago opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?

Enhancement to `DataTable` pinning mechanism. 

The current behavior is to set `position: sticky;` and `left: 0` to all pinned columns, leading to the reported case in #5130 where pinned columns get on top of each other.

This fix is based on filtering which columns have the `pin` property and then calculating the position offset each pinned column should have. This takes pinned column order into consideration and only considers pinned column widths, as users are not limited to pin a sequence of columns on the left side (i.e. users can pin the first and third columns, skipping the second).

#### Where should the reviewer start?

`src/js/components/DataTable/DataTable.js`

#### What testing has been done on this PR?

Manual testing as this involves `getBoundingClientRect`, but existing tests can confirm DataTable still works as intended.

#### How should this be manually tested?

Introduced a new story (`Visualizations/DataTable/Multiple pins` ) based on the provided example on the issue.

#### Any background context you want to provide?

#### What are the relevant issues?

Closes #5130 

#### Screenshots (if appropriate)

![Peek 2021-04-12 01-19](https://user-images.githubusercontent.com/47114840/114340124-38942a80-9b2d-11eb-83d1-255392b56bb6.gif)

#### Do the grommet docs need to be updated?

No.

#### Should this PR be mentioned in the release notes?

#### Is this change backwards compatible or is it a breaking change?

I'd consider it backward compatible as it doesn't affect overall usage. This will only affect cases where multiple pinned columns were set.

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -273,6 +273,40 @@ const DataTable = ({
+  const pinnedColumns = columns
+    .map(
+      (pinnedColumn, index) =>
+        pinnedColumn.pin && { property: pinnedColumn.property, id: index },
```
Do we need to return `id` at all? I think we can just leverage `index` on line `289`

taysea replied:
We could also probably just return `pinnedColumn.property` and create an array of strings instead of an array of objects with just one key.

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -273,6 +273,40 @@ const DataTable = ({
+    const firstRowCells = headerRef?.current?.firstChild?.cells || [];
+
+    if (firstRowCells.length !== 0) {
+      pinnedColumns.forEach(({ property, id }, index) => {
```
```suggestion
      pinnedColumns.forEach(({ property }, index) => {
```
Think we don't need `id` and can just use `index` in this function.

taysea commented on src/js/components/DataTable/stories/MultiplePin.js:
```diff
@@ -0,0 +1,331 @@
+      </Box>
+      <Box height="medium" overflow="auto">
+        <DataTable
+          data={data}
```
```suggestion
          background={{ pinned: 'background' }}
          data={data}
```
Small suggestion but just adding a background color to the pinned columns so when the data scrolls behind it doesn't become unreadable.

taysea commented on src/js/components/DataTable/stories/MultiplePin.js:
```diff
@@ -0,0 +1,331 @@
+              header: 'Id',
+              primary: true,
+              render: datum => datum.id.slice(datum.id.length - 5),
+              pin: size === 'small',
```
```suggestion
              pin: true,
```
Wondering if it'd be better to set the pin to `true` and then add a fixed width to the containing Box (`width="large"`) so people consuming the example won't have to resize their window to see the functionality.

taysea replied:
Would need adjusting on the two columns below this also.

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -273,6 +273,40 @@ const DataTable = ({
+    if (firstRowCells.length !== 0) {
+      pinnedColumns.forEach(({ property, id }, index) => {
+        if (firstRowCells[id]) {
+          pinnedPropertiesOrder.push(property);
```
If pinnedColumns becomes an array of strings, I think it would match `pinnedPropertiesOrder` and we wouldn't need this additional array.

taysea reviewed (commented):
Nice work here. Dropped some initial comments on a couple places where we might be able to simplify the `getPinnedOffset` function

leossantiago commented:
This took longer than expected.

@taysea just pushed a commit addressing some refactors to the Multiple Pins story, and also reworked the multiple pinning logic. As I mentioned previously, this also covers cases where the pinned columns aren't after each other.

While reworking it, I discovered a bug related to datatable not computing pinnedOffset properly after the first render in some situations, making the pinned columns override each other just as they used to behave. Fixed it by wrapping the logic into an `useEffect` hook.

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -173,6 +176,40 @@ const DataTable = ({
+    };
+
+    setPinnedOffset(getPinnedOffset());
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Can you add a comment about why we need to ignore exhaustive-deps warning? If dependency array is not desired, can we leave the `[]` out entirely?
```
useEffect(() => {...});
```

leossantiago replied:
We can try a workaround on that, but just removing the dependency array is leading to exhaustive-deps warnings. Probably because of line 209: `setPinnedOffset(getPinnedOffset());`

![image](https://user-images.githubusercontent.com/47114840/115761118-5e64cf00-a378-11eb-8b0f-42af835cff1b.png)


We can also set `columns` as a dependency. Might be clever considering we want this to update if columns are updated as well. What do you think?

taysea replied:
Following up from offline discussion. Recommending we try `useMemo` for `pinnedOffset` instead to potentially overcome the 'maximum depths' issue.

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -173,6 +176,40 @@ const DataTable = ({
+        .filter(n => n);
+
+      const pinnedColumnsProperties = {};
+      const firstRowCells = headerRef?.current?.firstChild?.cells || [];
```
Just for my own understanding, is the `?` syntax the equivalent of saying the following?
```
headerRef && headerRef.current && headerRef.current.firstChild ...
```

leossantiago replied:
Pretty much that. That was introduced as code was breaking when trying to access data from `headerRef.current` while it was undefined. Just realized this is kinda gone once we moved the logic to an useEffect hook. I'd suggest we go with one of those two approaches, in that case:

`headerRef.current.firstChild.cells || []`
or
`headerRef.current.firstChild.cells ?? []`

taysea commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -152,7 +152,41 @@ const DataTable = ({
+        }
+      });
+      return pinnedColumnsProperties;
+    }
```
```suggestion
    }
    return undefined;
```
Would it make sense to return undefined if we didn't set anything for `pinnedColumnsProperties` in this function? Effectively just reseting the state value to its default? If so, I think we could remove the eslint-disable comment

ericsoderberghp commented on src/js/components/DataTable/DataTable.js:
```diff
@@ -151,6 +151,43 @@ const DataTable = ({
+      .filter(n => n);
+
+    const pinnedColumnsProperties = {};
+    const firstRowCells = headerRef.current.firstChild.cells || [];
```
This is feeling like DataTable is reaching inside Header a bit too much. I'm wondering if we should move this inside Header and Cell, perhaps something like:
`<Header onWidths={([n,n,n]) => {}} />` and `<Cell onWidth={(width) => {}} />`
?

leossantiago replied:
I'm always open to suggestions, but I'm not sure if I'm following this one, sorry 😄 can you go a little further on that?

ericsoderberghp replied:
So, rather than have DataTable know that the Header's first child contains cells, which is knowing more about the Header than DataTable should, have the DataTable pass an `onWidths` to Header, indicating that the DataTable wants to be told the Header cell widths. When `onWidths` is passed to Header, Header passes an `onWidth` to each Cell. When a Cell has an `onWidth`, it reports its width, possible via a `useLayoutEffect()`. Header merges these into an array that it passes back up to DataTable.

The motivation is to not have DataTable know about the internals of the DOM structure of Header, to separate concerns. I am thinking this is just a few lines in each file. But, if going this route appears to be too complicated, I'm willing to consider other approaches.

ericsoderberghp reviewed (commented):
I like the general idea of calculating the header cell widths and then passing in the appropriate offset. I had one suggestion on how to structure the calculations.

leossantiago commented:
Updated the logic to separate responsibilities between cells, header, and DataTable.  However, just noticed there are some conflicts caused by #5177 . I'll address that ASAP.

ericsoderberghp reviewed (commented):
Just a few naming suggestions. Functionality looks good.

leossantiago commented:
Done! Let me know if you have any other suggestions 😄
